exercises/ubsan/solution/undefinedBehaviour.cpp

```diff
@@ -0,0 +1,60 @@
+#include <iostream>
+#include <limits>
+
+// You don't need to change any code in these structs:
+struct Base {
+  virtual void print() = 0;
+  virtual ~Base() {}
+};
+struct Derived1 : public Base {
+  void print() override {
+    std::cout << "Derived1::print()\n";
+  }
+};
+struct Derived2 : public Base {
+  void print() override {
+    std::cout << "Derived2::print()\n";
+  }
+};
+
+
+/**
+ * ***************
+ *  Instructions:
+ * ***************
+ *
+ * Compile and run this program using a compiler invocation like
+ * <compiler> -g -std=c++17 -o undefinedBehaviour undefinedBehaviour.cpp
+ *
+ * You might see warnings depending on the compiler, but on most platforms the program runs without observable issues.
+ * Smart compilers will warn with most of these bugs, but the compilers can easily be deceived by hiding a nullptr in
+ * a variable, or passing it as a function argument or similar.
+ *
+ * Since UBSan does runtime checks, it will catch these errors even if they are obscured.
+ * Recompile with "-fsanitize=undefined", and observe that almost every second line contains a serious bug.
+ * Try to understand what's wrong.
+ */
+
+int runTests() {
+  int arr[] = {1, 2, 3, 4};
+  std::cout << "arr[4]=" << arr[4] << "\n";             // Array overrun
+
+  unsigned short s = 1;
+  std::cout << "s << 33=" << (s << 33) << "\n";         // We cannot shift a type with 32 bits by 33 bits.
```

Comment:
And even less a 16 bits one !

Reply:
Interestingly (@sponce), it didn't trigger when I shifted by 17 bits. So I read up on cppreference, and it turns out the integral promotion is applied before shifting. The right-hand side is `int`, so it was correct ... 🙂

Reply:
indeed, that's interesting. But then I would use an unsigned int as I do not want to have to explain these details here.

Reply:
OK, good suggestion. I merged in the mean, time, but I will quickly edit the file in-place now.